tuf/ngclient/updater.py

```diff
@@ -122,12 +131,80 @@ def refresh(self) -> None:
             RepositoryError: Metadata failed to verify in some way
             DownloadError: Download of a metadata file failed in some way
         """
+        try:
+            version_bytes = self._load_local_metadata("spec_version")
+            self._spec_version = json.loads(version_bytes.decode("utf-8"))[
+                "version"
+            ]
+
+        except OSError as e:
+            if e.errno != errno.ENOENT:
+                raise
+            self._spec_version = "1"
+
+        repository_versions_and_paths = self._get_repository_versions()
+
+        repository_versions = [
+            i["version"] for i in repository_versions_and_paths
+        ]
+
+        # Updating self.spec_version
+        spec_version, message = _get_spec_version(
+            repository_versions, self._spec_version, self._supported_versions
+        )
+
+        if message:
+            logger.warning(message)
+        self._spec_version = spec_version
+
+        ordered_version_paths = []
 
-        self._load_root()
+        def filter_fn(a):
+            return a["version"] == version and version <= int(spec_version)
+
+        for version in sorted(repository_versions):
+            path = list(
+                filter(
+                    filter_fn,
+                    repository_versions_and_paths,
+                )
+            )
+            if len(path) > 0:
+                ordered_version_paths.append(path[0]["path"])
+
+        for i in repository_versions_and_paths:
+            if i["version"] == spec_version:
+                self._spec_version_dir = i["path"]
+
+        # persist the version number
+        self._persist_metadata(
+            "spec_version",
+            json.dumps({"version": spec_version}).encode("utf-8"),
+        )
+
+        self._load_root(ordered_version_paths)
         self._load_timestamp()
         self._load_snapshot()
         self._load_targets(Targets.type, Root.type)
 
+    def _get_repository_versions(self) -> List[str]:
+        """Returns a list of all the repository versions and paths."""
+
+        try:
+            url = f"{self._metadata_base_url}supported-versions.json"
+
+            target_bytes = self._fetcher.download_bytes(
+                url, self.config.supported_versions_max_length
+            )
+            repository_versions = json.loads(target_bytes)
```

Comment:
I would love to pull the parsing/types/etc. out into a `SupportedVersions` type that could be tested independently

Reply:
That's probably a more natural home for e.g. `_get_spec_versions`

Reply:
I also thought about moving some of this into the metadata api, especially if we finalize the decision to sign this file in theupdateframework/taps#158